Ea 2470 improved validation

SrdjanStevanetic opened the pull request.

SrdjanStevanetic committed:
contemporary commit (not tested)

SrdjanStevanetic committed:
validation of the entity fields (without testing)

SrdjanStevanetic committed:
validation of the entity fields tested

Commit:
Merge remote-tracking branch 'origin/develop' into
EA-2470-improved-validation

Conflicts:
	entity-management-web/src/integration-test/java/eu/europeana/entitymanagement/testutils/BaseMvcTestUtils.java
	entity-management-web/src/integration-test/java/eu/europeana/entitymanagement/web/EntityRecordServiceIT.java
	entity-management-web/src/test/java/eu/europeana/entitymanagement/validation/EntityFieldsValidatorTest.java

gsergiu committed:
Merge branch 'develop' into EA-2470-improved-validation

Commit:
#EA-2470 organize imports

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+		boolean returnValue=true;
+		
+		List<Field> entityFields = new ArrayList<>();
+		EntityUtils.getAllFields(entityFields, entity.getClass());
```
getAllFields method should return the list of fields and instantiate the array list itself

SrdjanStevanetic replied:
Improved.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+				 * the validation rules are implemented here
+				 */
+				//for the Text field type, check if it is a string and not a reference, and check the language codes
+				if (EntityFieldsTypes.valueOf(fieldName).getFieldType().equals("Text") || EntityFieldsTypes.valueOf(fieldName).getFieldType().equals("Keyword")) {
```
please use the constants from EntityFieldTypes instead of hardcoded value for TEST, URI

SrdjanStevanetic replied:
Improved.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+					returnValue = validateTextOrKeywordField(context, field, fieldValue, returnValue);
+				}
+				//for the URI field type, check if it is a reference
+				else if (EntityFieldsTypes.valueOf(fieldName).getFieldType().equals("URI") || EntityFieldsTypes.valueOf(fieldName).getFieldType().equals("Date") || EntityFieldsTypes.valueOf(fieldName).getFieldType().equals("Email")) {
```
EntityFieldsTypes.valueOf(fieldName).getFieldType() is used repetitively, please create a method in EntityFieldTyype.getType(String fieldName)

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+    }
+
+    @Override
+	public boolean isValid(Entity entity, ConstraintValidatorContext context) {		
```
the method should return the list of constrains and not a boolean value

SrdjanStevanetic replied:
Please note that this is the overriden boolean javax.validation.ConstraintValidator.isValid(T value, ConstraintValidatorContext context) method, which is called automatically when the validator annotation is found. The list of constraints is saved during the execution and can be retrived as we did in the test, by using the validate() function.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+				}				
+				//check the leading and trailing spaces from the String fields
+				else if (fieldType.isAssignableFrom(String.class)) {					
+					returnValue = validateStringField(context, field, fieldValue, returnValue);
```
the returnValue will be true or false depending if the last verified field is valid or not, regardless if the previous fields are invalid or not

SrdjanStevanetic replied:
Improved.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+    }
+    }
+    
+    boolean validateEmailOrDateOrURIField(ConstraintValidatorContext context, Field field, Object fieldValue, boolean returnValue) {
```
validation of Email value should be separated from validation of URIs

SrdjanStevanetic replied:
Improved.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+    }
+    
+    boolean checkEmailFormat(ConstraintValidatorContext context, Field field, Object fieldValue, boolean returnValue) {
+        String emailRegex = "^[a-zA-Z0-9_+&*-]+(?:\\."+
```
please use the EmailValidator fro apache commons: https://commons.apache.org/proper/commons-validator/apidocs/org/apache/commons/validator/routines/EmailValidator.html

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+			addConstraint(returnValue, context, "The entity field: "+field.getName()+" is of type: "+EntityFieldsTypes.valueOf(field.getName()).getFieldType()+" and it's element: "+fieldValue.toString()+" does not have the URI form.");
+			returnValue = false;
+		}
+    	else if (!shouldBeURI && EntityUtils.isUri(fieldValue.toString())) {
```
this is not a requirement in the ticket EA-2420, please remove https://europeana.atlassian.net/browse/EA-2420

SrdjanStevanetic replied:
I thought this is this requirement: Text: check if it is a string (not a reference). The same is for the Keyword.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+			Object[] fieldValueArray = (Object[])fieldValue;    			
+			if (fieldValueArray.length==0) return returnValue;   			
+			for (Object fieldValueElem : fieldValueArray) {
+				returnValue = checkURIFormat(context, field, fieldValueElem, returnValue, false);
```
verification of URI format must be performed only on fields that are specified as being URIs

SrdjanStevanetic replied:
This is done from the same reason as for the comment above. If we need to change it, just let me know.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+    
+    boolean validateNonMultilingualTextOrKeywordField(ConstraintValidatorContext context, Field field, Object fieldValue, boolean returnValue) {
+    	//if the field is an array
+		if (field.getType().isArray()) {
```
the fields were recently changed to List<String> instead of String[], in any case, the same method should be used for validation of string arrays and lists, by simply using Arrays.asList(String[]). Please change to this behaviour in all methods that handle string arrays

SrdjanStevanetic replied:
Improved.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+        return returnValue;
+    }
+    
+    boolean validateEmailOrDateOrURIArrayField(ConstraintValidatorContext context, Field field, Object fieldValue, boolean returnValue) {
```
please separate the validation of Email, Date and URI, there is no field which has the type EmailDateOrURI

SrdjanStevanetic replied:
Done.

gsergiu commented on entity-management-definitions/src/main/java/eu/europeana/entitymanagement/normalization/EntityFieldsValidator.java:
```diff
@@ -0,0 +1,299 @@
+    	return returnValue;
+    }
+   
+    boolean validateTextOrKeywordField (ConstraintValidatorContext context, Field field, Object fieldValue, boolean returnValue) {
```
please refactor to use validators for individual fields types, please do not mix the validators

SrdjanStevanetic replied:
I divided the validators, only for the Text and Keyword types I used the same function.

gsergiu reviewed (changes requested):
Please perform the required changes

SrdjanStevanetic committed:
1) improved entity fields validation tests; 2) corrected maven
dependency for the javax.xml.stream.XMLStreamWriter because of the
conflict with the system library

SrdjanStevanetic committed:
improved the validation of the entity fields

SrdjanStevanetic committed:
Merge branch 'develop' into EA-2470-improved-validation

Conflicts:
	entity-management-web/src/integration-test/java/eu/europeana/entitymanagement/web/EntityAdminControllerIT.java

Commit:
#EA-2420 code improvements

Commit:
#EA-2420 validate only fields with validation rules
